lib/spack/spack/test/spec_semantics.py

```diff
@@ -243,6 +245,123 @@ def test_satisfies_matching_variant(self):
+        # An abstract spec can instead be constrained
+        assert a.constrain(b)
+
+        a = Spec('multivalue_variant foo="bar,baz"', concrete=True)
+        spec_str = 'multivalue_variant foo="bar,baz,quux"'
+        b = Spec(spec_str)
+        assert not a.satisfies(b)
+        assert not a.satisfies(spec_str)
+        # A concrete spec cannot be constrained further
+        with pytest.raises(UnsatisfiableSpecError):
+            a.constrain(b)
+
+        a = Spec('multivalue_variant foo="bar,baz"')
+        spec_str = 'multivalue_variant foo="bar,baz,quux"'
+        b = Spec(spec_str)
+        assert not a.satisfies(b)
+        assert not a.satisfies(spec_str)
+        # An abstract spec can instead be constrained
+        assert a.constrain(b)
+        # ...but will fail during concretization if there are
+        # values in the variant that are not allowed
+        with pytest.raises(InvalidVariantValueError):
+            a.concretize()
+
+        # This time we'll try to set a single-valued variant
+        a = Spec('multivalue_variant fee="bar"')
+        spec_str = 'multivalue_variant fee="baz"'
+        b = Spec(spec_str)
+        assert not a.satisfies(b)
+        assert not a.satisfies(spec_str)
+        # A variant cannot be parsed as single-valued until we try to
+        # concretize. This means that we can constrain the variant above
+        assert a.constrain(b)
+        # ...but will fail during concretization if there are
+        # multiple values set
+        with pytest.raises(MultipleValuesInExclusiveVariantError):
+            a.concretize()
+
+        # FIXME: remove after having checked the correctness of the semantics
+        # check_unsatisfiable('multivalue_variant foo="bar,baz"',
+        #                     'multivalue_variant foo="bar,baz,quux"',
+        #                     concrete=True)
+        # check_unsatisfiable('multivalue_variant foo="bar,baz"',
+        #                     'multivalue_variant foo="bar,baz,quux"',
+        #                     concrete=True)
+
+        # but succeed for abstract ones (b/c they COULD satisfy the
+        # constraint if constrained)
+        # check_satisfies('multivalue_variant foo="bar"',
+        #                 'multivalue_variant foo="bar,baz"')
+
+        # check_satisfies('multivalue_variant foo="bar,baz"',
+        #                 'multivalue_variant foo="bar,baz,quux"')
+
+    def test_unsatisfiable_variant_types(self):
+        # These should fail due to incompatible types
+        check_unsatisfiable('multivalue_variant +foo',
+                            'multivalue_variant foo="bar"')
+
+        check_unsatisfiable('multivalue_variant ~foo',
+                            'multivalue_variant foo="bar"')
+
+        check_unsatisfiable('multivalue_variant foo="bar"',
+                            'multivalue_variant +foo')
+
+        check_unsatisfiable('multivalue_variant foo="bar"',
+                            'multivalue_variant ~foo')
```

Comment:
Is the test-case:
```python
def test_satisfies_matching_variant(self):
        check_satisfies('mpich+foo', 'mpich+foo')
        check_satisfies('mpich~foo', 'mpich~foo')
        check_satisfies('mpich foo=1', 'mpich foo=1')

        # confirm that synonymous syntax works correctly
        check_satisfies('mpich+foo', 'mpich foo=True')
        check_satisfies('mpich foo=true', 'mpich+foo')
        check_satisfies('mpich~foo', 'mpich foo=FALSE')
        check_satisfies('mpich foo=False', 'mpich~foo')
```
enough to check this? 

What I did for `BoolValuedVariant`s is to reserve (case-insensitive) "true" and "false". So `foo=TrUe` is the same as `+foo`.